drivers/display/display_st7735r.c

```diff
@@ -0,0 +1,461 @@
+			 (void *) write_data_start,
+			 desc->width * ST7735R_PIXEL_SIZE * write_h);
+
+	tx_bufs.buffers = &tx_buf;
+	tx_bufs.count = 1;
+
+	write_data_start += (desc->pitch * ST7735R_PIXEL_SIZE);
+	for (write_cnt = 1U; write_cnt < nbr_of_writes; ++write_cnt) {
+		tx_buf.buf = (void *)write_data_start;
+		tx_buf.len = desc->width * ST7735R_PIXEL_SIZE * write_h;
+		spi_write(data->spi_dev, &data->spi_config, &tx_bufs);
+		write_data_start += (desc->pitch * ST7735R_PIXEL_SIZE);
+	}
+
+	return 0;
+}
+
+static void *st7735r_get_framebuffer(const struct device *dev)
+{
+	return NULL;
+}
+
+static int st7735r_set_brightness(const struct device *dev,
+			   const uint8_t brightness)
+{
+	return -ENOTSUP;
+}
+
+static int st7735r_set_contrast(const struct device *dev,
+			 const uint8_t contrast)
+{
+	return -ENOTSUP;
+}
+
+static void st7735r_get_capabilities(const struct device *dev,
+			      struct display_capabilities *capabilities)
+{
+	struct st7735r_data *data = (struct st7735r_data *)dev->data;
+
+	memset(capabilities, 0, sizeof(struct display_capabilities));
+	capabilities->x_resolution = data->width;
+	capabilities->y_resolution = data->height;
+
+#ifdef CONFIG_ST7735R_RGB565
+	capabilities->supported_pixel_formats = PIXEL_FORMAT_RGB_565;
+	capabilities->current_pixel_format = PIXEL_FORMAT_RGB_565;
+#endif
+#ifdef CONFIG_ST7735R_BGR565
+	capabilities->supported_pixel_formats = PIXEL_FORMAT_BGR_565;
+	capabilities->current_pixel_format = PIXEL_FORMAT_BGR_565;
+#endif
+	capabilities->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+}
+
+static int st7735r_set_pixel_format(const struct device *dev,
+			     const enum display_pixel_format pixel_format)
+{
+#if defined CONFIG_ST7735R_RGB565
+	if (pixel_format == PIXEL_FORMAT_RGB_565) {
+#elif defined CONFIG_ST7735R_BGR565
+	if (pixel_format == PIXEL_FORMAT_BGR_565) {
+#endif
+		return 0;
+	}
+	LOG_ERR("Pixel format change not implemented");
+	return -ENOTSUP;
+}
+
+static int st7735r_set_orientation(const struct device *dev,
+			    const enum display_orientation orientation)
+{
+	if (orientation == DISPLAY_ORIENTATION_NORMAL) {
+		return 0;
+	}
+	LOG_ERR("Changing display orientation not implemented");
+	return -ENOTSUP;
+}
+
+static void st7735r_lcd_init(struct st7735r_data *p_st7735r)
+{
+	uint8_t tmp;
+
+	uint8_t tmparr[6];
+
+	st7735r_set_lcd_margins(p_st7735r, p_st7735r->x_offset,
+				p_st7735r->y_offset);
+
+	/* Frame Rate Control in Normal Mode, default value */
+	tmparr[0] = 1; tmparr[1] = 0x2c; tmparr[2] = 0x2d;
+	st7735r_transmit(p_st7735r, ST7735R_CMD_FRMCTR1, tmparr, 3);
+
+	/* Frame Rate Control in Idle Mode, default value */
+	tmparr[0] = 1; tmparr[1] = 0x2c; tmparr[2] = 0x2d;
+	st7735r_transmit(p_st7735r, ST7735R_CMD_FRMCTR2, tmparr, 3);
+
+	/* Frame Rate Control in Normal Mode, default value */
+	tmparr[0] = 1; tmparr[1] = 0x2c; tmparr[2] = 0x2d;
+	tmparr[3] = 1; tmparr[4] = 0x2c; tmparr[5] = 0x2d;
+	st7735r_transmit(p_st7735r, ST7735R_CMD_FRMCTR3, tmparr, 6);
```

Comment:
FRMCTR1, FRMCTR2, FRMCTR3, PWCTR[2...5], RASET,  CASET, VMCTR1 values should be obtained from device tree, like ST7735R_CMD_MADCTL.

Reply:
All parameters added